src/gsym/resolver.rs

```diff
@@ -151,76 +207,96 @@ impl SymResolver for GsymResolver<'_> {
+
+            if let Some(inline_info) = inline_info {
+                let inline_stack = inline_info.inline_stack(addr as u64);
+                let () = inlined.reserve(inline_stack.len());
+
+                for frame in inline_stack {
+                    let name = self
                         .ctx
-                        .get_str(finfo.filename as usize)
+                        .get_str(frame.name as usize)
+                        .and_then(|s| s.to_str())
                         .ok_or_invalid_data(|| {
-                            format!("failed to retrieve file name string @ {}", finfo.filename)
+                            format!("failed to read string table entry at offset {}", frame.name)
                         })?;
-                    return Ok(Some(AddrCodeInfo {
-                        dir: Path::new(dir),
-                        file,
-                        line: Some(lntab_row.file_line),
-                        column: None,
-                    }))
+
+                    let code_info = if let Some(file) = frame.call_file {
+                        let code_info = self.query_frame_code_info(file, frame.call_line)?;
+                        Some(code_info)
+                    } else {
+                        None
+                    };
+                    let () = inlined.push((name, code_info));
                 }
-                INFO_TYPE_INLINE_INFO => (),
-                typ => {
-                    warn!("encountered unknown info type: {typ}; ignoring...");
-                    continue
+
+                if !inlined.is_empty() {
+                    let (name, _code_info) = inlined.remove(0);
+                    let mut prev = Option::<usize>::None;
+
+                    for i in 0..inlined.len() {
+                        if let Some(prev_i) = prev {
+                            let inlined = inlined.as_mut_ptr();
+                            // TODO: Use `slice::get_many_mut` once it is stable.
+                            // SAFETY: `i` and `prev_i` are different so
+                            //         we are creating exclusive
+                            //         references to disjunct region of
+                            //         memory. It is an invariant that
+                            //         both are within bounds of the
+                            //         `symbols` slice.
+                            let first = unsafe { &mut *inlined.add(i) };
+                            let second = unsafe { &mut *inlined.add(prev_i) };
+                            let () = swap(&mut first.1, &mut second.1);
+                        } else {
+                            direct_name = Some(name);
+                            if let Some(code_info) = &mut inlined[i].1 {
+                                let () = swap(code_info, &mut line_tab_info);
+                            }
+                        }
+
+                        prev = Some(i);
```

Comment:
this two-step construction seems unnecessary convoluted, tbh. Wouldn't it be simpler to construct final `inlined` vector on the first pass? It seems like you could avoid unsafe bits as well? E.g.,  first, from inline_stack[0] you extract name and keep it as "pending name", while code_info goes into sym.code_info (store code_info from line_table for later separately for now). Then on each item in inline_step[1..len), take code_info and combine it with last pending name, reassign pending name to be name you just extracted.

After the whole loop, take last pending name and combine it with code_info from line_table as last inlined element.

Reply:
Thanks for the comment. Okay, will take another look.

Reply:
https://github.com/libbpf/blazesym/pull/328